### ISSUE TITLE
Tech discovery: Plans and unknowns around Q&A reusability in Drupal

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#9579]
## Description

This issue is for back-end/Drupal devs to get onboarded to the technical aspects of making Q&A content reusable – specifically within the Resources & Support product. 

Among things to understand:
- Understand what has already been determined from a content type and architecture perspective
- Understand what has not been decided yet
- And raise questions that need further exploration

**Document user stories that you can pursue in upcoming sprints.**

## Acceptance Criteria
- [ ] Met with engineers with context (Lynn, Steve)
- [x] Read epic
- [x] Explored code base and data examples

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [ ] `⭐️ Sitewide CMS`
- [x] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

**Comment by chri5tia** [comment]
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/9530#issuecomment-1165148955

![Screen Shot 2022-06-23 at 8 10 41 PM](https://user-images.githubusercontent.com/22764938/175455837-3bf610d5-5741-4879-bc09-a41b09d78d63.png)

https://docs.google.com/spreadsheets/d/18DOxHt0t0ugT6eUiu9etyGdGi9jCqgthQndT6KAAS8E/edit#gid=0

What is needed:

- [ ] Establish order of operations (refine spreadsheet above) before writing plan
- [ ] Get stakeholder approval of plan before assigning the work
- [ ] Include here some screenshots and data examples, confirm the discovery work outlined in the epic (#8630) for where all the Q&A type of content exists

**Comment by chri5tia** [comment]
Making more of the existing Question and Answer text reusable is going to require:

- [ ] remove duplicates
- [ ] re-enter the removed duplicates that were on existing "pages" from the resuable 
- [ ] client input as to which question and answer text should be made resuable
- [ ] continue to support both the resuable and nonresuable methods of input
- [ ] sync features (robustness?) between the resuable and non-resuable Q&A text forms, find out which features (such as rich. text)
- [ ] reusable entity select modal will need a search or filter capability

**Comment by chri5tia** [comment]
Architectural changes would affect multiple teams/platform wide.

**Comment by jilladams** [comment]
Front-end ticket with notes! https://github.com/department-of-veterans-affairs/va.gov-cms/issues/9530

**Comment by jilladams** [comment]
The lost note I had in draft and didn't publish yesterday!!: 

Dave mid-sprint checkin notes:
We should schedule a follow up conversation (preferably when Wes is back) with Beth & Danielle from the Sitewide Content team about the discoverability / migration question from CMS perspective, including CMS folks. I'll work on scheduling. 

https://agilesix.slack.com/archives/D03KE0BALHX/p1656369261545649

**Comment by jilladams** [comment]
@chri5tia I think we'll meet next Thurs 7/7 to close loop. Could you add to the research here to talk to @swirtSJW and do some digging re: what options are available for autocomplete in the Place Q&A modal on FAQ page nodes? (Where it exists today.) It will be helpful to understand any technical limitations / performance concerns before we propose that to client.

If autocomplete on title is a nonstarter, the fall back would be some sort of taxonomy, I'm guessing, so wouldn't hurt to consider the next best option.

**Comment by chri5tia** [comment]
https://app.mural.co/t/vagov6717/m/vagov6717/1656088606354/882d5358ec3463a39230521d787f0015233e777c?sender=ff5a5026-64ad-4b35-9e26-5ec47e2da32c

**Comment by chri5tia** [comment]
EOD/Sprint note: got my local dev in a functional and exploratory state and will have this updated early next week.

**Comment by chri5tia** [comment]
#8630 context and findings previously

**Comment by chri5tia** [comment]
Item: Add Q&A Groups to Resource & Support Creation Form

Currently when adding a main content block, if you do a search for "q", there are no Q&A sections available:
![Screen Shot 2022-07-07 at 1 03 12 AM](https://user-images.githubusercontent.com/22764938/177740346-53c41a1f-5df9-47c9-8f60-7d065762efad.png)

- [ ] Edit content type `support_resource_detail_page` field `field_content` to include paragraph type `q_a_group`.

![Screen Shot 2022-07-07 at 1 04 49 AM](https://user-images.githubusercontent.com/22764938/177740125-7007373c-b654-4a82-99b2-8d82722a8caf.png)

When complete, Q&A Groups comes up for an option.

**Comment by chri5tia** [comment]
Item: Discoverability of Q&As in modal

The Q&A Group paragraph type (`q_a_groups`) is reusable. Its "Q&As" field (`field_q_as`) form display uses "Entity Browser - Table" with the Q&A Entity Browser, which opens a modal which allows editors to select 1 or multiple Q&A items by title. It also provides a filter:

![Screen Shot 2022-07-07 at 1 44 17 AM](https://user-images.githubusercontent.com/22764938/177731948-d8836abd-6338-47a0-b85b-a76b33ffebcb.png)
![Screen Shot 2022-07-07 at 1 45 14 AM](https://user-images.githubusercontent.com/22764938/177731953-ede3461e-5d09-43ce-a049-9c126150a5e4.png)

Note: [Entity Browser Table](https://www.drupal.org/project/entity_browser_table) is a contrib module dependent on entity_browser.

We could add an autocomplete field type to the modal by editing the paragraph `q_a_group` field widget type to "Autocomplete"
![Screen Shot 2022-07-07 at 2 14 38 AM](https://user-images.githubusercontent.com/22764938/177738523-2f31b917-25a0-41dd-bc4e-941986de671d.png)

![Screen Shot 2022-07-07 at 2 17 40 AM](https://user-images.githubusercontent.com/22764938/177741458-ecc79ea8-a79d-437a-9711-f5acd1739ef2.png)

Note: This option removes the modal but the autocomplete method may serve the same purpose.

- [ ] How important is it to keep the modal?

To add the autocomplete field within the entity brower, we may have to write a contrib module:
https://www.drupal.org/project/entity_browser/issues/3136427

**Comment by chri5tia** [comment]
Further exploring discoverability in modal.

The entity browser types are views exposed forms (machine name `content_entity_browsers`)
at admin/structure/views/view/content_entity_browsers

![Screen Shot 2022-07-07 at 2 50 16 AM](https://user-images.githubusercontent.com/22764938/177745630-9c4c73b4-6cfe-4819-8744-52d38051618d.png)
![Screen Shot 2022-07-07 at 2 50 07 AM](https://user-images.githubusercontent.com/22764938/177745636-c48455db-8c2a-42a9-b43d-791952baf570.png)

A filter can be added to the Q&A entity browser as an exposed title/keyword field for "contains" operator.

![Screen Shot 2022-07-07 at 2 59 45 AM](https://user-images.githubusercontent.com/22764938/177747972-21f24979-02ec-41cf-ba16-87c2578173d2.png)

![Screen Shot 2022-07-07 at 3 01 23 AM](https://user-images.githubusercontent.com/22764938/177748285-21cf2eef-60d3-4941-9efe-627ff3a7473e.png)

![Screen Shot 2022-07-07 at 3 01 40 AM](https://user-images.githubusercontent.com/22764938/177748304-b30d84e4-e0ad-4c32-9ee2-e3c4836c8f44.png)

- [x] Ready to commit

**Comment by chri5tia** [comment]
Some other tech discovery findings and facts, building on the discovery work done in #8630 

**Content types that use Q&A**

1. Campaign Landing Pages
2. Benefits Detail Page
3. Centralized Content
4. VAMC Detail
5. Q&A - Single
6. FAQ Page

**1. Campaign Landing Pages**

   - Non-resuable Q&A text is entered via field `field_clp_faq_paragraphs` where paragraph type "Q&A" is made available. "Q&A Groups" can be made available as well.
   - nodes: 10, 8 published
   - machine name: `campaign_landing_page`
  
![Screen Shot 2022-07-06 at 7 44 44 PM](https://user-images.githubusercontent.com/22764938/177752924-209015a9-d668-4ac0-8a29-8ac70180728d.png)

UX: 
  
  - Go to Content > Add content > Campaign Landing Page. Then Expand FAQs > Enable this page segment > Add Q&A.

  - Rich Text WYSIWYG available.

![Screen Shot 2022-07-06 at 7 49 30 PM](https://user-images.githubusercontent.com/22764938/177754297-c858eaa9-d63e-4cfd-85ad-345cad289fab.png)

![Screen Shot 2022-07-06 at 7 50 45 PM](https://user-images.githubusercontent.com/22764938/177754295-badb77ea-4676-4ce9-9eb7-cbdbe2ac9fe1.png)


**Benefits Detail Page**

   - Non-reusable Q&A text is entered via field `field_content_block` where paragraph type "Q&A Sections" is made available.
   - nodes: 349, 294 published
   - machine name: `page`

UX:

  - Go to Content > Add content > Benefits Detail Page. Then "Add main section" > search for "Q" in the autocomplete box and create new text.

- not on "Add content" fly-out menu

![Screen Shot 2022-07-06 at 8 04 16 PM](https://user-images.githubusercontent.com/22764938/177755806-9c5fd5f3-ef22-480f-b9be-5202a9392f52.png)

**Centralized Content**

 - Non-reusable Q&A text via field `field_content_block` and either paragraph type "Q&A" or "Q&A Section"
 - nodes: 8, 6 published
 - machine name: `centralized_content`

UX:

- Add content > Centralized Content > Add content block > Search and select Q&A

**VAMC Detail**

- Non-reusable Q&A text via field `field_content_block` and paragraph type "Q&A Section". 
- nodes: 5714, 5231 published
- machine name: `health_care_region_detail_page`

UX: 

- Go to content > add content > VAMC detail page > featured content > add Q&A

![Screen Shot 2022-07-06 at 9 35 44 PM](https://user-images.githubusercontent.com/22764938/177757703-76621c7d-97fa-4437-baea-775572b6c7ce.png)

![Screen Shot 2022-07-06 at 9 46 39 PM](https://user-images.githubusercontent.com/22764938/177757700-f20ed25b-0d9b-4b9b-bed0-27e5e27383f7.png)

**Q&A - Single**

- an individual node, content type that can be a standalone page or referenced. 
- nodes: 202, 148 published
- machine: `q_a`
- fields: `answer` and `title` (question)

UX:

- Add content > Q&A - Single > Enter question as the title and then answer.
- Rich text available.

**FAQ Page**
  
- reuseable Q&A text via field `field_q_a_group` and paragraph type "Q&A Groups".
- nodes: 29, 23 published
- machine name: `faq_multiple_q_a`

UX:

- Add content > FAQ Page > Place Q&As
- can't be sorted
- can select non-published
- can select one or many

**Comment by chri5tia** [comment]
TL;DR
- keyword/search filter added to modal for selecting reusable questions
- reusable Q&A content block option added to Resources content form
- reusable Q&A content block option can be added to other content types while preserving the option for non-reusable that is already being used

**Comment by chri5tia** [comment]
Updated Q&A entity browser view / Q&A Group modal to include a preview of the answer text as well as exclude archived Q&As.

![Screen Shot 2022-07-07 at 1 47 17 PM](https://user-images.githubusercontent.com/22764938/177868566-527c3992-616f-47d6-94e9-3df249e456e4.png)

**Comment by chri5tia** [comment]
Closing in lieu of PR #9737 and related #9745 as next steps.
